src/core/thread/mesh_forwarder_ftd.cpp

```diff
@@ -235,20 +235,31 @@ void MeshForwarder::UpdateIndirectMessages(void)
     }
 }
 
-otError MeshForwarder::EvictIndirectMessage(void)
+otError MeshForwarder::EvictMessage(uint8_t aPriority)
 {
-    otError error = OT_ERROR_NOT_FOUND;
+    otError  error = OT_ERROR_NOT_FOUND;
+    Message *message;
 
-    for (Message *message = mSendQueue.GetHead(); message; message = message->GetNext())
-    {
-        if (!message->IsChildPending())
-        {
-            continue;
-        }
+    VerifyOrExit((message = mSendQueue.GetTail()) != NULL);
 
+    if (message->GetPriority() < aPriority)
+    {
         RemoveMessage(*message);
         ExitNow(error = OT_ERROR_NONE);
     }
+    else if (message->GetPriority() == aPriority)
+    {
+        for (message = mSendQueue.GetHead(); message; message = message->GetNext())
```

Comment:
If tail message's priority is same as `aPriority` (meaning there is no message in queue with lower priority), we get here. 
In that case, we start from `mSendQueue.GetHead()` which will start from highest priority message in queue and remove the first one that has `IsChildPending` as true (destined for sleepy child). 

I know we did similar logic in original implementation (without considering the priority). But thinking more about this, I think it may be better if we first try to remove a message at same priority level as `aPriority`.

How about if we start the search from `GetHeadForPriority(aPriority)` (to start from message with same priority as `aPriority`).  If that fails and we cannot find any message at this priority level (destined to sleepy child) to remove, we can then search the next priority level (keep increasing the priority level to see if we can find a message to remove). 

Thoughts?

Reply:
You are right. It is better to remove the message at the same priority level firstly. I will update the code.